module-jni: compiler diagnostics not captured in diagnostic object but rather output on stderr

```
2025-02-24 10:39:55.050537 professor-e.prague.qoretechnologies.com:53235 T1735 [ERROR]: System exception: QoreJavaCompiler.java:354 (Java): JNI-ERROR: org.qore.jni.compiler.QoreJavaCompilerException
: Compilation failed. (compiling {com.qoretechnologies.qorus.regression.test-java-service: "711 bytes"}: no compiler diagnostic output available) (compiling service "test-java-service" v1.0)
```

actual stderr output:
```
test-java-service.java:9: error: class TestJavaService is public, should be declared in a file named TestJavaService.java
public class TestJavaService extends QorusService {
       ^
test-java-service.java:9: warning: [deprecation] QorusService in com.qoretechnologies.qorus.service has been deprecated
public class TestJavaService extends QorusService {
                                     ^
1 error
1 warning
```

--- a/docs/mainpage.dox.tmpl
+++ b/docs/mainpage.dox.tmpl
@@ -868,6 +868,9 @@ set_save_object_callback(callback);
       (<a href="https://github.com/qoretechnologies/qore/issues/5034">issue 5034</a>)
     - fixed a CMake bug where inner class inc file dependencies were not tracked correctly due to a typo
       (<a href="https://github.com/qoretechnologies/qore/issues/5034">issue 5034</a>)
+    - fixed a bug where Java compiler diagnostics were output to stderr instead of being captured in the
+      exception message
+      (<a href="https://github.com/qoretechnologies/qore/issues/4925">issue 4925</a>)
 
     @subsection jni_2_4_0 jni Module Version 2.4.0
     - added the <a href="../../ExcepDataProvider/html/index.html">ExcepDataProvider</a> module to allow for reading

--- a/src/java/org/qore/jni/compiler/QoreJavaCompiler.java
+++ b/src/java/org/qore/jni/compiler/QoreJavaCompiler.java
@@ -6,6 +6,7 @@ import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.io.StringWriter;
 
 import java.nio.charset.StandardCharsets;
 
@@ -326,6 +327,13 @@ public class QoreJavaCompiler<T> {
             final Map<String, CharSequence> classes,
             final DiagnosticCollector<JavaFileObject> diagnosticsList)
             throws QoreJavaCompilerException {
+        // Always use a DiagnosticCollector to capture compiler diagnostics
+        DiagnosticCollector<JavaFileObject> localDiagnostics = diagnosticsList;
+        if (localDiagnostics == null) {
+            localDiagnostics = new DiagnosticCollector<JavaFileObject>();
+        }
+        diagnostics = localDiagnostics;
+
         List<JavaFileObject> sources = new ArrayList<JavaFileObject>();
         for (Entry<String, CharSequence> entry : classes.entrySet()) {
             String qualifiedClassName = entry.getKey();
@@ -345,13 +353,20 @@ public class QoreJavaCompiler<T> {
                         className + JAVA_EXTENSION, source);
             }
         }
+        // Use a StringWriter to capture any compiler output that would otherwise go to stderr
+        StringWriter compilerOutput = new StringWriter();
         try {
             // Get a CompliationTask from the compiler and compile the sources
-            final CompilationTask task = compiler.getTask(null, javaFileManager, diagnostics,
+            final CompilationTask task = compiler.getTask(compilerOutput, javaFileManager, localDiagnostics,
                     options, null, sources);
             final Boolean result = task.call();
             if (result == null || !result) {
-                throw new QoreJavaCompilerException("Compilation failed.", classes.keySet(), diagnostics);
+                String outputStr = compilerOutput.toString();
+                String message = "Compilation failed.";
+                if (!outputStr.isEmpty()) {
+                    message += " Compiler output: " + outputStr;
+                }
+                throw new QoreJavaCompilerException(message, classes.keySet(), localDiagnostics);
             }
             try {
                 // For each class name in the output map, get compiled class and file data and put it in the output map
@@ -366,11 +381,11 @@ public class QoreJavaCompiler<T> {
                 }
                 return compiled;
             } catch (ClassNotFoundException e) {
-                throw new QoreJavaCompilerException(classes.keySet(), e, diagnostics);
+                throw new QoreJavaCompilerException(classes.keySet(), e, localDiagnostics);
             } catch (IllegalArgumentException e) {
-                throw new QoreJavaCompilerException(classes.keySet(), e, diagnostics);
+                throw new QoreJavaCompilerException(classes.keySet(), e, localDiagnostics);
             } catch (SecurityException e) {
-                throw new QoreJavaCompilerException(classes.keySet(), e, diagnostics);
+                throw new QoreJavaCompilerException(classes.keySet(), e, localDiagnostics);
             }
         } finally {
             classLoader.clearCompilationCache();

